pkg/cluster/config/encoding/scheme.go

```diff
@@ -43,59 +49,172 @@ func init() {
+func unmarshalConfig(contents []byte) (*config.Config, error) {
+	// parses yamlDocuments separated by --- directives
+	yamlDocuments, err := splitYAMLDocuments(contents)
 	if err != nil {
-		return nil, errors.Wrap(err, "decoding failure")
+		return nil, err
 	}
 
-	// apply defaults
-	Scheme.Default(cfg)
-
-	// converts to internal cfg
-	var internalCfg = &config.Config{}
-	Scheme.Convert(cfg, internalCfg, nil)
+	// checks if v1alpha1 config is present in the yamlDocuments
+	var v1alpha1Config = false
+	for _, doc := range yamlDocuments {
+		if doc.GroupVersionKind.GroupVersion() == v1alpha1.SchemeGroupVersion {
+			v1alpha1Config = true
+		}
+	}
 
-	return internalCfg, nil
-}
+	// if using v1alpha1Config,
+	if v1alpha1Config {
+		// only one yaml document is supported
+		if len(yamlDocuments) > 1 {
+			return nil, fmt.Errorf("if using v1alpha1 config, only one yaml document should be provided")
+		}
+
+		// decode the yaml into a config object
+		var v1alpha1Cfg = &v1alpha1.Config{}
+		_, _, err := Codecs.UniversalDecoder().Decode(yamlDocuments[0].Contents, nil, v1alpha1Cfg)
+		if err != nil {
+			return nil, fmt.Errorf("decoding failure: %v", err)
+		}
+
+		// apply defaults
+		Scheme.Default(v1alpha1Cfg)
+
+		// converts to internal cfg. this will give Config with a single node with control-plane role
+		// NB. we are using custom conversion due to the complexity of conversion from v1alpha1 config
+		var internalCfg = &config.Config{}
+		v1alpha1Cfg.Convert(internalCfg)
+
+		return internalCfg, nil
+	}
 
-// Load reads the file at path and attempts to convert into a `kind` Config; the file
-// can be one of the different API versions defined in scheme.
-// If path == "" then the default config is returned
-func Load(path string) (*config.Config, error) {
-	if path == "" {
-		return newDefaultedConfig(), nil
+	// Otherwise it is a multi node, v1alpha2 config
+	var cfg = &config.Config{}
+
+	// Process all the yamlDocuments
+	for _, doc := range yamlDocuments {
+		// decode data into a Node object
+		var v1alpha2node = &v1alpha2.Node{}
+		_, _, err := Codecs.UniversalDecoder().Decode(doc.Contents, nil, v1alpha2node)
+		if err != nil {
+			return nil, fmt.Errorf("decoding failure: %v", err)
+		}
+
+		// apply defaults
+		Scheme.Default(v1alpha2node)
+
+		// converts to internal cfg
+		var internalNode = &config.Node{}
+		Scheme.Convert(v1alpha2node, internalNode, nil)
+
+		// adds the node to the internal config
+		// in case replicas shuold be generated for the nodes, Add will takes care.
+		if err := cfg.Add(internalNode); err != nil {
+			return nil, err
+		}
 	}
 
-	// read in file
-	contents, err := ioutil.ReadFile(path)
-	if err != nil {
-		return nil, err
+	return cfg, nil
+}
+
+// splitYAMLDocuments divides a yaml file into yamlDocuments,
+// and identifies each document with its own GroupVersionKind
+func splitYAMLDocuments(contents []byte) (yamlDocuments []yamlDocument, err error) {
```

Comment:
we really need a unified location for this type of utility.
this is pretty much duplicating kubeadm at this point. :\

and once the ComponentConfig KEP stuff is in place we need to revise all locations that bake their own utility.